File: 4ourth/Python/main.py
from math import log10
from itertools import count


from math import log10


class Handler:
    def __init__(self, a, b, c, d):
        self.a = a
        self.b = b
        self.c = c
        self.d = d
        self.__result = None

    def calculate(self, debug=False):
        # [lg(4b-c)a]/[b+c/d-1]
        if debug:
            print("\nПочаток обчислення")
        if debug:
            print(f'[lg(4*{self.b}-{self.c})*{self.a}]/[{self.b}+{self.c}/{self.d}-1] = ', end='')
        first = 4 * self.b
        if debug:
            print(f'[lg({first}-{self.c})*{self.a}]/[{self.b}+{self.c}/{self.d}-1] = ', end='')
        second = first - self.c
        if debug:
            print(f'[lg({second})*{self.a}]/[{self.b}+{self.c}/{self.d}-1] = ', end='')
        try:
            third = log10(second)
            if debug:
                print(f'[{third})*{self.a}]/[{self.b}+{self.c}/{self.d}-1] = ', end='')
        except ValueError:
            raise ValueError(f"Базою логарифму не може бути {second}")
        fourth = third * self.a
        if debug:
            print(f'[{fourth}]/[{self.b}+{self.c}/{self.d}-1] = ', end='')
        try:
            fifth = self.c/self.d
            if debug:
                print(f'[{fourth}]/[{self.b}+{fifth}-1] = ', end='')
        except ZeroDivisionError:
            raise ZeroDivisionError("d не може дорівнювати 0")
        sixth = self.b + fifth
        if debug:
            print(f'[{fourth}]/[{sixth}-1] = ', end='')
        seventh = sixth - 1
        if debug:
            print(f'{fourth}/{seventh} = ', end='')
        try:
            self.__result = fourth / seventh
            if debug:
                print(f'{self.__result}')
        except ZeroDivisionError:
            raise ZeroDivisionError("b+c/d-1 не може дорівнювати 0")
        return first, second, third, fourth, fifth, sixth, seventh

    def getResult(self):
        if self.__result is None:
            raise NotImplementedError("Значення ще не обчислене")
        else:
            return self.__result

    def logger(self):
        with open('logs/python.log.csv', "a+") as log:
            calc = ();
            try:
                calc = self.calculate()
            except ZeroDivisionError:
                log.write('e0')
            except ValueError:
                log.write('eL')
            except:
                log.write('e?')
            try:
                log.write(f'{self.getResult()}, {", ".join(str(i) for i in calc)}\r\n')
            except NotImplementedError:
                log.write('eI\r\n')


if __name__ == '__main__':
    with open("start_values.tsv", "r") as values:
        values = values.readlines()
        values = list(map(lambda x: x.split('\t'), values))

    from itertools import count

    with open("logs/python.log.csv", "w+") as log:
        log.write('Result,First,Second,Third,Fourth,Fifth,Sixth,Seventh\r\n')
    for data, iteration in zip(values, count()):
        handler = Handler(
            float(data[0]),
            float(data[1]),
            float(data[2]),
            float(data[3]),
        )
        handler.logger()

File: 4ourth/C++/main.cpp
#include <iostream>
#include <cmath>
#include <stdexcept>
#include <fstream>

class Handler
{
private:
    double result;
public:
    double getResult() {
        if (isnan(result)) throw std::domain_error("NotImplementedError: Значення ще не обчислене");
        else return result;
    }
    double a, b, c, d;
    Handler(double a, double b, double c, double d) {
        this->a = a;
        this->b = b;
        this->c = c;
        this->d = d;
        this->result = nan("0");
    }
    double * calculate(bool debug=false) {
        double first, second, third, fourth, fifth, sixth, seventh;
        static double ret[7];
        if (debug) std::cout << "\nПочаток обчислення" << std::endl;
        if (debug) std::cout << "[lg(4*" << b << "-" << c << ")*" << a << "]/[" << b << "+" << c << "/" << d << "-1] = ";
        first = 4 * b;
        ret[0] = first;
        if (debug) std::cout << "[lg(" << first << "-" << c << ")*" << a << "]/[" << b << "+" << c << "/" << d << "-1] = ";
        second = first - c;
        ret[1] = second;
        if (debug) std::cout << "[lg(" << second << ")*" << a << "]/[" << b << "+" << c << "/" << d << "-1] = ";
        if (second<=0) throw std::invalid_argument("ValueError: Базою логарифму не може бути <= 0");
        else {
            third = log10(second);
            ret[2] = third;
            if (debug) std::cout << "[" << third << "*" << a << "]/[" << b << "+" << c << "/" << d << "-1] = ";
        }
        fourth = third * a;
        ret[3] = fourth;
        if (debug) std::cout << "[" << fourth << "]/[" << b << "+" << c << "/" << d << "-1] = ";
        if (d==0) throw std::logic_error("ZeroDivisionError: d не може дорівнювати 0");
        else {
            fifth = c/d;
            ret[4] = fifth;
            if (debug) std::cout << "[" << fourth << "]/[" << b << "+" << fifth << "-1] = ";
        }
        sixth = b + fifth;
        ret[5] = sixth;
        if (debug) std::cout << "[" << fourth << "]/[" << sixth << "-1] = ";
        seventh = sixth - 1;
        ret[6] = seventh;
        if (debug) std::cout << "[" << fourth << "]/[" << seventh << "] = ";
        if (seventh==0) throw std::logic_error("ZeroDivisionError: b+c/d-1 не може дорівнювати 0");
        if (debug) std::cout << fourth / seventh << std::endl;
        this->result = fourth / seventh;
        return ret;
    }

    void logger(){
        std::ofstream log;
        double *calc;
        log.open("logs/c++.log.csv", std::ios_base::app);
        try {
            calc = calculate();
        }
        catch (const std::invalid_argument &e) {
            log << "eL";
        }
        catch (const std::logic_error &e) {
            log << "e0";
        }
        catch (const char *msg) {
            log << "e?";
        }
        try {
            log<<getResult()<< ", "<<calc[0]<<", "<<calc[1]<<", "<<calc[2]<<", "<<calc[3]<<", "<<calc[4]<<", "<<calc[5]<<", "<<calc[6]<<"\r\n";
        }
        catch (const std::domain_error &e) {
            log << "eI\r\n";
        }
        log.close();
    }
};

int main() {
    std::ifstream values;
    values.open("start_values.tsv");
    double a, b, c, d;
    std::ofstream log_f;
    log_f.open("logs/c++.log.csv", std::ios_base::out | std::ios_base::trunc);
    log_f << "Result,First,Second,Third,Fourth,Fifth,Sixth,Seventh\r\n";
    log_f.close();
    while (values >> a >> b >> c >> d){
        Handler handler(a, b, c, d);
        handler.logger();
    }
}
